File: Python/Concurrency/BoundedQueueDemo.py
import threading
import time
from collections import deque
from concurrent.futures import ThreadPoolExecutor
from typing import Deque, Generic, TypeVar

T = TypeVar("T")


class BoundedQueue(Generic[T]):
    def __init__(self, capacity: int) -> None:
        self._queue: Deque[T] = deque()
        self._capacity = capacity
        self._condition = threading.Condition()

    def put(self, item: T) -> None:
        with self._condition:
            while len(self._queue) == self._capacity:  # full — wait for a taker
                print(f"{threading.current_thread().name} sees FULL "
                      f"({len(self._queue)}/{self._capacity}), waiting...")
                self._condition.wait()
            self._queue.append(item)
            print(f"{threading.current_thread().name} put: {item} [size={len(self._queue)}]")
            self._condition.notify_all()  # wake possible waiting takers (and putters)

    def take(self) -> T:
        with self._condition:
            while not self._queue:  # empty — wait for a putter
                print(f"{threading.current_thread().name} sees EMPTY, waiting...")
                self._condition.wait()
            item = self._queue.popleft()
            print(f"{threading.current_thread().name} took: {item} [size={len(self._queue)}]")
            self._condition.notify_all()
            return item


def producer(queue: "BoundedQueue[str]") -> None:
    for i in range(5):
        queue.put(f"{threading.current_thread().name}-item{i}")
        time.sleep(0.03)


def consumer(queue: "BoundedQueue[str]") -> None:
    for _ in range(5):
        queue.take()
        time.sleep(0.06)


def main() -> None:
    queue: BoundedQueue[str] = BoundedQueue(3)  # capacity 3, not 1

    with ThreadPoolExecutor(max_workers=4) as pool:
        pool.submit(producer, queue)
        pool.submit(producer, queue)
        pool.submit(consumer, queue)
        pool.submit(consumer, queue)


if __name__ == "__main__":
    main()

File: JavaScript/Concurrency/BoundedQueueDemo.js
'use strict';

class BoundedQueue {
    constructor(capacity) {
        this.queue = [];
        this.capacity = capacity;
        this.waiters = [];
    }

    async put(item, name) {
        while (this.queue.length === this.capacity) {  // full — wait for a taker
            console.log(`${name} sees FULL (${this.queue.length}/${this.capacity}), waiting...`);
            await new Promise((resolve) => this.waiters.push(resolve));
        }
        this.queue.push(item);
        console.log(`${name} put: ${item} [size=${this.queue.length}]`);
        this._notifyAll();  // wake possible waiting takers (and putters)
    }

    async take(name) {
        while (this.queue.length === 0) {  // empty — wait for a putter
            console.log(`${name} sees EMPTY, waiting...`);
            await new Promise((resolve) => this.waiters.push(resolve));
        }
        const item = this.queue.shift();
        console.log(`${name} took: ${item} [size=${this.queue.length}]`);
        this._notifyAll();
        return item;
    }

    _notifyAll() {
        const waiters = this.waiters;
        this.waiters = [];
        waiters.forEach((resolve) => resolve());
    }
}

function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

async function producer(queue, name) {
    for (let i = 0; i < 5; i++) {
        await queue.put(`${name}-item${i}`, name);
        await sleep(30);
    }
}

async function consumer(queue, name) {
    for (let i = 0; i < 5; i++) {
        await queue.take(name);
        await sleep(60);
    }
}

async function main() {
    const queue = new BoundedQueue(3);  // capacity 3, not 1

    await Promise.all([
        producer(queue, 'producer-1'),
        producer(queue, 'producer-2'),
        consumer(queue, 'consumer-1'),
        consumer(queue, 'consumer-2'),
    ]);
}

main();
